File: Medium/3634_minimum_removals_to_balance_array/solution.py
class Solution:
    def minRemoval(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if n == 1: return 0
        nums.sort()
        # max <= min * k
        # return the number of removals
        sw = 0 # track the longest sliding window
        l = 0
        for r in range(1, n):
            while nums[r] > nums[l] * k:
                l+=1
            length = r-l+1
            if sw < length: 
                sw = length
        return n - sw
        

File: Medium/3634_minimum_removals_to_balance_array/solution.cpp
class Solution {
public:
    int minRemoval(vector<int>& nums, int k) {
        int n = nums.size();
        if (n==1) return 0;
        std::sort(nums.begin(), nums.end());
        int length = 0;
        int sw = 0;
        int l = 0;
        for (int r=1; r<n; r++) {
            while (nums[r] > ((long long)nums[l]*k)) l++; // (1LL*nums[l]*k) also works
            length = r-l+1;
            sw = max(length, sw);
        }
        return n - sw;
    }
};
